nixos/modules/services/backup/postgresql-wal-receiver.nix

```diff
@@ -0,0 +1,210 @@
+          to ensure that timely feedback is sent to the server.
+        '';
+      };
+
+      compress = mkOption {
+        type = types.ints.between 0 9;
+        default = 0;
+        description = ''
+          Enables gzip compression of write-ahead logs, and specifies the compression level
+          (<literal>0</literal> through <literal>9</literal>, <literal>0</literal> being no compression and <literal>9</literal> being best compression).
+          The suffix <literal>.gz</literal> will automatically be added to all filenames.
+
+          This option requires PostgreSQL >= 10.
+        '';
+      };
+
+      connection = mkOption {
+        type = types.str;
+        example = "postgresql://user@somehost";
+        description = ''
+          Specifies parameters used to connect to the server, as a connection string.
+          See <link xlink:href="https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING">Section 34.1.1 of the PostgreSQL manual</link> for more information.
+
+          Because <command>pg_receivewal</command> doesn't connect to any particular database in the cluster,
+          database name in the connection string will be ignored.
+        '';
+      };
+
+      environment = mkOption {
+        type = with types; attrsOf str;
+        default = { };
+        example = literalExample ''
+          {
+            PGPASSFILE = "/private/passfile";
+            PGSSLMODE = "require";
+          }
+        '';
+        description = ''
+          Environment variables passed to the service.
+          Usable parameters are listed in <link xlink:href="https://www.postgresql.org/docs/current/libpq-envars.html">Section 34.14 of the PostgreSQL manual</link>.
+        '';
+      };
+    };
+  };
+
+in {
+  options = {
+    services.postgresqlWalReceiver = {
+      receivers = mkOption {
+        type = with types; attrsOf (submodule receiverSubmodule);
+        default = { };
+        example = literalExample ''
+          {
+            main = {
+              postgresqlPackage = pkgs.postgresql_9_6;
+              directory = /mnt/pg_wal/main/;
+              slot = "main_wal_receiver";
+              connection = "postgresql://user@somehost";
+            };
+          }
+        '';
+        description = ''
+          PostgreSQL WAL receivers.
+          Stream write-ahead logs from a PostgreSQL server using <command>pg_receivewal</command> (formerly <command>pg_receivexlog</command>).
+          See <link xlink:href="https://www.postgresql.org/docs/current/app-pgreceivewal.html">the man page</link> for more information.
+        '';
+      };
+    };
+  };
+
+  config = let
+    receivers = config.services.postgresqlWalReceiver.receivers;
+  in mkIf (receivers != { }) {
+    users = {
+      groups.postgresql-wal-receiver = { };
+      users.postgresql-wal-receiver = {
```

Comment:
[StackOverflow](https://serverfault.com/questions/294121/what-is-the-maximum-username-length-on-current-gnu-linux-systems) says better stick to short usernames. Something like `walreceiver` would be fine, what do you think?

And also, if it is possible to implement master-slave replication using this module, maybe the running user should be configurable? In test you explicitly do `chown -R postgres:postgres`, but it may not strictly required.

Reply:
> StackOverflow says better stick to short usernames. Something like walreceiver would be fine, what do you think?

Indeed. I've modified the user and group accordingly. The user's description should be enough.

> And also, if it is possible to implement master-slave replication using this module, maybe the running user should be configurable? 

This program and module solely for proper WAL archiving.
Making the user and group configurable might be useful but I cannot think of a real use-case.

> In test you explicitly do chown -R postgres:postgres, but it may not strictly required.

Indeed, but on a real setup, the WAL receiver would run on a separate machine and the restoration would have used a copy of the log.